### PR TITLE
Making settleAuctionAtomically and precalculateSellAmountSum exclusive

### PR DESCRIPTION
**Pull request by josojo**

### CODE REVIEW

**Comment by fedgiac** [comment]
I'd consider removing this feature completely.
Restriction 1: the auctioneer who enables this feature cannot rely on the fact that this settlement method can be used in the future. Anyone can call `precalculateSellAmount` or `settleAuction` on solution submission, which de facto disables the atomic settlement.
Restriction 2: atomic settlement can only be used when there aren't many orders (`settleAuctionAtomically` must fit a block). I didn't do any test, but I imagine settling 100 orders is already impossible.

I understand the use case of closing the auction with an order that balances the price with a counter order from another DEX, like on Uniswap. This is in practice a free arbitrage opportunity for whoever manages to settle first and I'm not fully convinced of the value of this use case.

If we really want to go through with atomic settlements anyway, I'd just allow whoever settles to create an order smaller than `interimSumBidAmount`. This is basically equivalent to these changes except that `precalculateSellAmount` does not immediately disable the feature.

**Comment by josojo** [comment]
>Restriction 1: the auctioneer who enables this feature cannot rely on the fact that this settlement method can be used in the future. Anyone can call precalculateSellAmount or settleAuction on solution submission, which de facto disables the atomic settlement.

I disagree. If there is an arbitrage opportunity, then people calling the `settleAuctionAtomically` will always outbid other settlement transactions, as they can afford to pay a higher gas price.

>Restriction 2: atomic settlement can only be used when there aren't many orders (settleAuctionAtomically must fit a block). I didn't do any test, but I imagine settling 100 orders is already impossible.

I think it is more about 1000 orders. And by setting a minBuyAmountPerOrder, the auctioneer can enforce that the auction can be settled atomically.

> If we really want to go through with atomic settlements anyway, I'd just allow whoever settles to create an order smaller than interimSumBidAmount. This is basically equivalent to these changes except that precalculateSellAmount does not immediately disable the feature.

I don't think that it will really make a difference in practice, as there is no incentive to call precalculateSellAmount without finishing the settlement. But I am fine with changing it to your and adams suggestion.

**Comment by fedgiac** [comment]
> I think it is more about 1000 orders. And by setting a minBuyAmountPerOrder, the auctioneer can enforce that the auction can be settled atomically.

I tested this statement, and indeed the contract can settle 1000 orders with less than 2M gas, nice!
<details><summary>test</summary>

Result:
```
  EasyAuction
    settleAuction
Gas used by settle for 1003 orders: 1864089
      ✓ simple version of e2e gas test (561304ms)


  1 passing (9m)

Done in 567.22s.
```

I slightly modified an existing test to accept an arbitrary amount of orders:
```
    it.only("simple version of e2e gas test", async () => {
      const initialAuctionOrder = {
        sellAmount: ethers.utils.parseEther("1"),
        buyAmount: ethers.utils.parseEther("1"),
        userId: BigNumber.from(0),
      };
      let sellOrders = [];
      const count = 1000;
      for (let i = 0; i < count; i++) {
        sellOrders.push({
          sellAmount: ethers.utils.parseEther("1").div(4).div(count),
          buyAmount: ethers.utils.parseEther("1").div(8).div(count).add(i),
          userId: BigNumber.from(0),
        });
      }
      sellOrders = sellOrders.concat([
        {
          sellAmount: ethers.utils.parseEther("1").div(4),
          buyAmount: ethers.utils.parseEther("1").div(12),
          userId: BigNumber.from(0),
        },
        {
          sellAmount: ethers.utils.parseEther("1").div(4),
          buyAmount: ethers.utils.parseEther("1").div(16),
          userId: BigNumber.from(0),
        },
        {
          sellAmount: ethers.utils.parseEther("1").div(4),
          buyAmount: ethers.utils.parseEther("1").div(20),
          userId: BigNumber.from(0),
        },
      ]);
      const {
        auctioningToken,
        biddingToken,
      } = await createTokensAndMintAndApprove(
        easyAuction,
        [user_1, user_2],
        hre,
      );

      const auctionId: BigNumber = await sendTxAndGetReturnValue(
        easyAuction,
        "initiateAuction(address,address,uint256,uint256,uint96,uint96,uint256,uint256,bool)",
        auctioningToken.address,
        biddingToken.address,
        60 * 60,
        60 * 60,
        initialAuctionOrder.sellAmount,
        initialAuctionOrder.buyAmount,
        1,
        0,
        false,
      );
      await placeOrders(easyAuction, sellOrders, auctionId, hre);

      await closeAuction(easyAuction, auctionId);
      const price = await calculateClearingPrice(easyAuction, auctionId);

      const call = await easyAuction.settleAuction(auctionId);
      const { gasUsed } = await call.wait();
      console.log(
        `Gas used by settle for ${count + 3} orders: ${gasUsed.toString()}`,
      );
      expect(price).to.eql(initialAuctionOrder);
      const auctionData = await easyAuction.auctionData(auctionId);
      expect(auctionData.clearingPriceOrder).to.equal(
        encodeOrder(reverseOrderPrice(initialAuctionOrder)),
      );
      expect(auctionData.volumeClearingPriceOrder).to.equal(
        initialAuctionOrder.sellAmount,
      );
      await claimFromAllOrders(easyAuction, auctionId, sellOrders);
    });
```
</details>

> I don't think that it will really make a difference in practice, as there is no incentive to call precalculateSellAmount without finishing the settlement. But I am fine with changing it to your and adams suggestion.

I can imagine somebody who wants to disrupt the auction "for the lulz" calling `precalculateSellAmount(id, 1)`.

My issue is that I don't understand who stands to gain from this feature. So far, it looks like it would only be the arbitrageur.

**Comment by josojo** [comment]
>My issue is that I don't understand who stands to gain from this feature. So far, it looks like it would only be the arbitrageur.

If the auction without the arbitrageur would only be partially filled, then at least the arbitrageur could increase the sell volume to 100%. I think this is benefital for DAOs auctioning of their tokens.

If the auction is fully filled, then the arbitrageur can improve the price for the auctioneer, though there is no guarantee for it. It could be the same price, though there is no disadvantage for the auctioneer.

>I can imagine somebody who wants to disrupt the auction "for the lulz" calling precalculateSellAmount(id, 1).

Not sure how real it is. But yeah. Let's change it to prevent the "attack."

**Comment by fedgiac** [comment]
> If the auction is fully filled, then the arbitrageur can improve the price for the auctioneer, though there is no guarantee for it. It could be the same price, though there is no disadvantage for the auctioneer.

The disadvantage is that it offers an option to manipulate the price on a silver platter. The auction is closed, yet after the fact at most one person can create an order with information that was not available when all other orders were created. For example, I can imagine a situation where someone with a large amount of funds didn't put an order to avoid raising the price and then scoops up most of the tokens at the end. My understanding of this project is that it wants to offer a fair platform for auctioning new tokens for both users and auctioneers, and allowing this option goes against this vision.

**Comment by josojo** [comment]
>My understanding of this project is that it wants to offer a fair platform for auctioning new tokens for both users and auctioneers, and allowing this option goes against this vision.

Yeah, for this vision, this option should neverr be considered.

Though, it might be that DAOs like to auction off some tokens that are already available and where the AMM slippage is too big to sell it directly. Here, this feature could be useful

**Comment by josojo** [comment]
> > I can imagine somebody who wants to disrupt the auction "for the lulz" calling precalculateSellAmount(id, 1).
> 
> Not sure how real it is. But yeah. Let's change it to prevent the "attack."

I think, a more effective way of blocking this attack is to not allow precalculateSellAmount calls, if the atomic option is enabled.

**Comment by fedgiac** [comment]
> I think, a more effective way of blocking this attack is to not allow precalculateSellAmount calls, if the atomic option is enabled.

But this is very dangerous, since a successful IDO could have more orders that they can settle in a single block, in which case all funds would be stuck.

I see why you wouldn't like the proposed approach however: calling `precalculateSellAmount` makes it impossible to change the price to one that is higher than the last price involved in the precalculation.

File: contracts/EasyAuction.sol [hunk 1]
```diff
@@ -367,6 +367,21 @@ contract EasyAuction is Ownable {
             auctionData[auctionId].isAtomicClosureAllowed,
             "not allowed to settle auction atomically"
         );
+        require(
+            _minBuyAmount.length == 1 && _sellAmount.length == 1,
+            "Only one order is allowed to place atomically"
```
Comment (fedgiac) [inline contracts/EasyAuction.sol]:
```suggestion
            "Only one order can be placed atomically"
```

File: test/contract/EasyAuction.spec.ts [hunk 2]
```diff
@@ -2391,6 +2391,177 @@ describe("EasyAuction", async () => {
...
+
+      await closeAuction(easyAuction, auctionId);
+      await expect(
+        easyAuction.settleAuctionAtomically(
+          auctionId,
+          [atomicSellOrders[0].sellAmount, atomicSellOrders[1].sellAmount],
+          [atomicSellOrders[0].buyAmount, atomicSellOrders[1].buyAmount],
+          [queueStartElement, queueStartElement],
+        ),
+      ).to.be.revertedWith("Only one order is allowed to place atomically");
```
Comment (fedgiac) [inline test/contract/EasyAuction.spec.ts]:
```suggestion
      ).to.be.revertedWith("Only one order can be placed atomically");
```

**Review by fedgiac — APPROVED** [review]

**Pull request merged**